compiler/stmt.py

```diff
@@ -32,15 +32,106 @@
+    "unicode_literals": (FUTURE_UNICODE_LITERALS, False),
+}
+
+# These future features are already in the language proper as of 2.6, so
+# importing them via __future__ has no effect.
+redundant_future_features = ["generators", "with_statement", "nested_scopes"]
+
+late_future = 'from __future__ imports must occur at the beginning of the file'
+
+
+def import_from_future(node):
+  """processes a future import statement, returning set of flags it defines."""
+  assert isinstance(node, ast.ImportFrom)
+  assert node.module == '__future__'
+  flags = 0
+  for alias in node.names:
+    name = alias.name
+    if name in future_features:
+      flag, implemented = future_features[name]
+      if not implemented:
+        msg = 'future feature {} not yet implemented by grumpy'.format(name)
+        raise util.ParseError(node, msg)
+      flags |= flag
+    elif name == 'braces':
+      raise util.ParseError(node, 'not a chance')
+    elif name not in redundant_future_features:
+      msg = 'future feature {} is not defined'.format(name)
+      raise util.ParseError(node, msg)
+  return flags
+
+
+class FutureFeatures(object):
+  def __init__(self):
+    self.parser_flags = 0
+    self.future_lineno = 0
+
+
+def visit_future(node):
+  """Accumulates a set of compiler flags for the compiler __future__ imports.
+
+  Returns an instance of FutureFeatures which encapsulates the flags and the
+  line number of the last valid future import parsed. A downstream parser can
+  use the latter to detect invalid future imports that appear too late in the
+  file.
+  """
+  # If this is the module node, do an initial pass through the module body's
+  # statements to detect future imports and process their directives (i.e.,
+  # set compiler flags), and detect ones that don't appear at the beginning of
+  # the file. The only things that can proceed a future statement are other
+  # future statements and/or a doc string.
+  assert isinstance(node, ast.Module)
+  ff = FutureFeatures()
+  done = False
+  found_docstring = False
+  for node in node.body:
+    if isinstance(node, ast.ImportFrom):
+      modname = node.module
+      if modname == '__future__':
+        if done:
+          raise util.ParseError(node, late_future)
+        ff.parser_flags |= import_from_future(node)
+        ff.future_lineno = node.lineno
+      else:
+        done = True
+    elif isinstance(node, ast.Expr) and not found_docstring:
+      e = node.value
+      if not isinstance(e, ast.Str): # pylint: disable=simplifiable-if-statement
+        done = True
+      else:
+        found_docstring = True
+    else:
+      done = True
+  return ff
+
+
 class StatementVisitor(ast.NodeVisitor):
   """Outputs Go statements to a Writer for the given Python nodes."""
 
   # pylint: disable=invalid-name,missing-docstring
 
   def __init__(self, block_):
     self.block = block_
+    self.future_features = self.block.future_features or FutureFeatures()
```

Comment:
The pattern I've used for accessing global state like this (and I'm not saying it's a good pattern, just an established one) is to put a property on Block that pulls the corresponding attribute from self._module_block. See for example [full_package_name](https://github.com/google/grumpy/blob/master/compiler/block.py#L83).

future_features seems like a similar kind of global state. Can we use that pattern instead?